### PR TITLE
Process.Unix: consider executable permission while searching PATH.

### LINKED ISSUE #29940: broken symlinks and non-executable files not ignored in FindProgramInPath()
I bumped into a difficult-to-debug problem when using `Process.Start()` on Linux, which ultimately was due to the fact that the `FindProgramInPath()` internal method used to look for a given executable file in each component of the `$PATH` environment variable does not ignore broken symlinks and non-executable files, the way Linux shells and other tools such as `which` do.

Consider the following scenario, where `/usr/local/bin/date` is a bad symlink which will be found first when searching `$PATH`; however, normal system utilities ignore the broken symlink and proceed to find and use `/bin/date`:

```
$ /bin/ls -lF /bin/date
-rwxr-xr-x 1 root root 100568 Jan 18  2018 /bin/date*
$ /bin/ls -lF /usr/local/bin/date
lrwxrwxrwx 1 root root 10 Jun 18 11:02 /usr/local/bin/date -> /tmp/dummy
$ printenv PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
$ which date
/bin/date
$ date
Tue Jun 18 17:32:10 EDT 2019
```

The following example .NET program, though, throws an exception:

```
using System;
using System.Diagnostics;

namespace Foo
{
    class Program
    {
        static void Main(string[] args)
        {
            ProcessStartInfo procInfo = new ProcessStartInfo("date");
            procInfo.WorkingDirectory = "/tmp";
            procInfo.UseShellExecute = false;
            procInfo.RedirectStandardOutput = true;
            procInfo.RedirectStandardError = true;

            using (Process proc = new Process())
            {
                proc.StartInfo = procInfo;
                proc.Start();
                Console.WriteLine($"out: {proc.StandardOutput.ReadToEnd()}");
                proc.WaitForExit();
            }
        }
    }
}
```

```
$ dotnet run
Unhandled Exception: System.ComponentModel.Win32Exception: No such file or directory
   at Interop.Sys.ForkAndExecProcess(String filename, String[] argv, String[] envp, String cwd, Boolean redirectStdin, Boolean redirectStdout, Boolean redirectStderr, Boolean setUser, UInt32 userId, UInt32 groupId, Int32& lpChildPid, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean shouldThrow)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at Foo.Program.Main(String[] args)
```

I believe this is due to the fact that the [`File.Exists()` check](https://github.com/dotnet/corefx/blob/800f1b3da5fe1519d333e7772ba98b2277564d4a/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L742) in `FindProgramInPath()` in `src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs` is insufficient to determine if the path is merely a symlink, or an actual executable file.

Note that the same example C# code above also fails if `/usr/local/bin/date` is a regular file which lacks an execute permission mode setting, e.g.:

```
$ sudo touch /usr/local/bin/date
$ ls -lF /usr/local/bin/date
-rw-r--r-- 1 root root 0 Jun 18 14:52 /usr/local/bin/date
$ which date
/bin/date
$ date
Tue Jun 18 17:32:10 EDT 2019
$ dotnet run
Unhandled Exception: System.ComponentModel.Win32Exception: Permission denied
   at Interop.Sys.ForkAndExecProcess(String filename, String[] argv, String[] envp, String cwd, Boolean redirectStdin, Boolean redirectStdout, Boolean redirectStderr, Boolean setUser, UInt32 userId, UInt32 groupId, Int32& lpChildPid, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean shouldThrow)
```

It would be ideal if, instead of using `File.Exists()`, these tests (and others used in [`ResolvePath()`](https://github.com/dotnet/corefx/blob/800f1b3da5fe1519d333e7772ba98b2277564d4a/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L682)) could instead perform `stat(2)` and `readlink(2)` calls to check whether the path corresponds to a valid executable file.

While this wouldn't prevent races with file deletion, etc. (which can happen with the existing code, too), under normal circumstances it would bring the .NET parsing of `$PATH` more in line with other system utilities.

And if that isn't possible, then at least having the code output the path used in the [`ForkAndExecProcess()`](https://github.com/dotnet/corefx/blob/800f1b3da5fe1519d333e7772ba98b2277564d4a/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L542) call would be superb, because otherwise it's very difficult and frustrating to ascertain why an executable which can be executed without problems from the command line can't be executed by a .NET program.

(In my case, I ended up tracking down the relevant code here, then cut-and-pasting it over into my program and debugging it piece by piece until I realized it was finding a broken symlink from earlier in `$PATH`.  There may be better ways to do that, but regardless, having an error message which output the path that was not found would be incredibly helpful!)

### PR DESCRIPTION
**Pull request by tmds**
Fixes https://github.com/dotnet/runtime/issues/29940

@stephentoub @carlossanlop @adamsitnik @jozkee ptal

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 1]
```diff
@@ -1954,7 +1954,7 @@
       <Link>Common\Interop\Unix\System.Native\Interop.GetCwd.cs</Link>
     </Compile>
     <Compile Include="$(CommonPath)Interop\Unix\System.Native\Interop.GetEGid.cs">
-      <Link>Common\Interop\Unix\System.Native\Interop.GetHostName.cs</Link>
+      <Link>Common\Interop\Unix\System.Native\Interop.GetEGid.cs</Link>
```
Comment (tmds) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:1957]: Fixes an unrelated typo.

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        internal static unsafe uint[]? GetGroups()
+        {
+            const int InitialGroupsLength =
+#if DEBUG
+                1;
+#else
+                64;
+#endif
+            Span<uint> groups = stackalloc uint[InitialGroupsLength];
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs:19]: I assume the chances someone actually has more than 64 groups is exceedingly small?

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 3]
```diff
@@ -752,6 +755,43 @@ private static string[] CreateEnvp(ProcessStartInfo psi)
...
+                return (permissions & Interop.Sys.Permissions.S_IXUSR) != 0;
+            }
+            // A group we're a member of owns the file.
+            else if (s_egid == fileinfo.Gid ||
+                    (s_groups != null && Array.BinarySearch(s_groups, fileinfo.Gid) >= 0))
+            {
+                return (permissions & Interop.Sys.Permissions.S_IXGRP) != 0;
+            }
+            // Other.
+            return (permissions & Interop.Sys.Permissions.S_IXOTH) != 0;
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:792]: Nit:
```suggestion
            if (s_euid == 0)
            {
                // We're root.
                return (permissions & Interop.Sys.Permissions.S_IXUGO) != 0;
            }

            if (s_euid == fileinfo.Uid)
            {
                // We own the file.
                return (permissions & Interop.Sys.Permissions.S_IXUSR) != 0;
            }
            
            if (s_egid == fileinfo.Gid ||
                (s_groups != null && Array.BinarySearch(s_groups, fileinfo.Gid) >= 0))
            {
                // A group we're a member of owns the file.
                return (permissions & Interop.Sys.Permissions.S_IXGRP) != 0;
            }

            // Other.
            return (permissions & Interop.Sys.Permissions.S_IXOTH) != 0;
```

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            do
+            {
+                int rv;
+                int ngroups = groups.Length;
+                fixed (uint* pGroups = groups)
+                {
+                    rv = Interop.Sys.GetGroups(ngroups, pGroups);
+                }
+                if (rv >= 0)
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs:28]: Nit:
```suggestion
                }

                if (rv >= 0)
```

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                {
+                    // increase buffer size
+                    groups = new uint[groups.Length * 2];
+                }
+                else
+                {
+                    // failure
+                    return null;
+                }
+            } while (true);
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs:43]: Nit:
```suggestion
            }
            while (true);
```

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+#endif
+            Span<uint> groups = stackalloc uint[InitialGroupsLength];
+            do
+            {
+                int rv;
+                int ngroups = groups.Length;
+                fixed (uint* pGroups = groups)
+                {
+                    rv = Interop.Sys.GetGroups(ngroups, pGroups);
+                }
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs:27]:
```suggestion
                int rv;
                fixed (uint* pGroups = groups)
                {
                    rv = Interop.Sys.GetGroups(groups.Length, pGroups);
                }

```

**Review by stephentoub — APPROVED** [review]
Thanks!

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs (hunk 2 above)
Comment (tmds) [reply src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetGroups.cs:19]: Yes, 64 is plenty.

**Comment by tmds** [comment]
CI failures are unrelated.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**